### ISSUE TITLE
DRC does not test clearance between segments/vias of the same net

### ISSUE THREAD

**Issue by marekr** [marekr/kicad-issue-test#7]
While routing a track, I found it was possible to place two consecutive vias under the required clearance.

At first, I thought this was a via to via issue, but this is not so, because if the vias do not belong to the same net, it is impossible to place them under the said clearance.

The same is true for track segments: you can create a track that passes by itself very closely (as in a coil loop), without triggering DRC errors.

Attached is a simple board illustrating what I think should trigger at least 3 DRC errors, even if it's the same net:

- Via to Via
- Via to segment
- segment to segment

All of this is due to these lines, saying basically that no DRC errors can be triggered by same net objects:

drc_clearance_test_functions.cpp L345

        // No problem if segments have the same net code:
        if( net_code_ref == track->GetNet() )
            continue;

I understand that these lines alleviate the need for testing connectivity at the same time as DRC (two segments physically linked together do not violate DRC, but I think two segments of the same net, when linked physically through a third segment, should be tested for DRC).

In other words, the fact that two segments/vias share the same net does not guaranty this cannot lead to fabrication problems.

Guillaume

Application: Pcbnew
Version: (2011-12-16 BZR 3308)-testing
Build: wxWidgets 2.8.11 (no debug,Unicode,compiler with C++ ABI 1002,GCC 4.4.4,wx containers,compatible with 2.4,compatible with 2.6)
Platform: Linux 2.6.32-131.12.1.el6.x86_64 x86_64, 64 bit, Little endian, wxGTK
Options: USE_PNG_BITMAPS=ON
         KICAD_GOST=OFF
         USE_WX_GRAPHICS_CONTEXT=OFF
         USE_WX_OVERLAY=OFF
         USE_BOOST_POLYGON_LIBRARY

**Comment by marekr** [comment]
(by gsimard)

**Comment by marekr** [comment]
(by dickelbeck)
I agree with you about via to via.

I would suggest we try and separate the discussion into 2 categories:

1) Maybe we can all agree that via to via is a bug?

2) Find a way to gather more opinion on :

a) via to track of same net.
b) track to track of same net.


By more opinion, I mean more people's opinions.

**Comment by marekr** [comment]
(by dickelbeck)
I agree with you about via to via.

I would suggest we try and separate the discussion into 3 categories

1) agree that via to via is a bug?

get more opinions and ideas (from more people) on :

2a) via to track of same net.

2b) track to track of same net.

**Comment by marekr** [comment]
(by gsimard)
1) Via-to-Via: Agreed, I'm working on it, I have fixed the offline DRC now I'm trying to add the online DRC for a new via while routing.

2a) and 2b) I agree we need more people commenting on this, but I think it's a fab issue, so it should trigger a DRC error, perhaps of a different kind (saying it's the same net).

**Comment by marekr** [comment]
(by bidulock)
Dick,

On Fri, 16 Dec 2011, Dick Hollenbeck wrote:

> I agree with you about via to via.
> 
> I would suggest we try and separate the discussion into 3 categories
> 
> 1) agree that via to via is a bug?
> 
> get more opinions and ideas (from more people) on :
> 
> 2a) via to track of same net.
> 
> 2b) track to track of same net.
> 

Add pads and zone into there as well.  There are design
rules for same-net copper to copper clearances: kicad just
doesn't have any of them.

--brian

-- 
Brian F. G. Bidulock    � The reasonable man adapts himself to the �
<email address hidden>    � world; the unreasonable one persists in  �
http://www.openss7.org/ � trying  to adapt the  world  to himself. �
                        � Therefore  all  progress  depends on the �
                        � unreasonable man. -- George Bernard Shaw �

**Comment by marekr** [comment]
(by gsimard)
This is a patch for proposition #1, ie.: DRC testing for same net via-to-via spacing.

Two things have been done:

- DRC testing is no longer skipped for same net objects IF the two objects are vias.
- Online DRC now also checks against the track being currently routed (it was testing for all board tracks, excluding the track being routed).

Because the DRC still skips same net objects (except for via-via), testing DRC against the currently routed track does not change behaviour of the online DRC, apart from solving the problem of being able to place two successive vias very closely while routing.

Guillaume
